services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryManager.java

```diff
@@ -0,0 +1,166 @@
+        future.whenComplete((r, t) -> {
+            if (t != null) {
+                failedUploads.add(DefaultFailedUpload.builder()
+                                                     .exception(t)
+                                                     .path(path)
+                                                     .build());
+            } else {
+                uploads.add(r);
+            }
+        });
+        return future;
+    }
+
+    private Stream<Path> listFiles(Path directory, UploadDirectoryRequest request) {
+
+        try {
+            boolean recursive = transferConfiguration.resolveUploadDirectoryRecursive(request);
+
+            if (!recursive) {
+                return Files.list(directory).filter(Files::isRegularFile);
+            }
+
+            boolean followSymbolicLinks = transferConfiguration.resolveUploadDirectoryFollowSymbolicLinks(request);
+
+            int maxDepth = transferConfiguration.resolveUploadDirectoryMaxDepth(request);
+
+            if (followSymbolicLinks) {
+                return Files.walk(directory, maxDepth, FileVisitOption.FOLLOW_LINKS)
+                            .peek(p -> log.trace(() -> "Processing path: " + p))
+                            .filter(Files::isRegularFile);
+            }
+
+            return Files.walk(directory, maxDepth)
+                        .peek(p -> log.trace(() -> "Processing path: " + p))
+                        .filter(path -> Files.isRegularFile(path, LinkOption.NOFOLLOW_LINKS));
+
+        } catch (IOException e) {
+            throw SdkClientException.create("Failed to list files under the provided directory", e);
+        }
+    }
+
+    private static String processPrefix(String prefix) {
+        if (StringUtils.isEmpty(prefix)) {
+            return "";
+        }
+        return prefix.endsWith("/") ? prefix : prefix + "/";
+    }
+
+    private static String getRelativePathName(int directoryNameCount, Path path) {
+        String relativePathName = path.subpath(directoryNameCount,
+                                               path.getNameCount()).toString();
+
+        // Replace "\" (Windows FS) with "/"
+        return relativePathName.replace('\\', '/');
+    }
+
+    private static UploadRequest constructUploadRequest(UploadDirectoryRequest uploadDirectoryRequest, int directoryNameCount,
+                                                        Path path) {
+        String prefix = processPrefix(uploadDirectoryRequest.prefix());
+        String relativePathName = getRelativePathName(directoryNameCount, path);
+        String key = prefix + relativePathName;
+
+        PutObjectRequest putObjectRequest = PutObjectRequest.builder()
+                                                            .bucket(uploadDirectoryRequest.bucket())
+                                                            .key(key)
+                                                            .build();
```

Comment:
Currently, we are just passing bucket and key and I'm thinking about introducing a transfer extension (similar to the core execution interceptor) for this to allow users to provide modify other parameters in `PutObjectRequest`.

Reply:
They could do this with an ExecutionInterceptor on the underlying S3 client, once we support configuring the underlying S3 client, right? 

I suppose a `putObjectRequestSupplier(Supplier<PutObjectRequest>)` on the UploadDirectoryRequest would be more friendly, especially if they want request-level parameter configuration. +1

Reply:
> They could do this with an ExecutionInterceptor on the underlying S3 client, once we support configuring the underlying S3 client, right?

Yes, that's right.

> I suppose a putObjectRequestSupplier(Supplier<PutObjectRequest>) on the UploadDirectoryRequest would be more friendly, especially if they want request-level parameter configuration. +1

+1. It would be much easier than what I proposed. I suppose we can use `Consumer<PutObjectRequest.Builder>` to make it clear that this is a modification.

Reply:
I agree that offering a way to hook into the resulting `PutObjectRequest` is crucial.

There are a few other use cases I can think of:
1. I want to be able to supply my own `Path`-filters. For example, I may want to ensure I exclude sensitive file extensions. E.g., a `Predicate<Path>`.
1. I want to be able to customize a `key` name. For example, I may want to force everything to lowercase. E.g., a configurable `Function<Path,String>`, or a `UnaryOperator<String>`.
1. I want to be able to transform file contents before they are uploaded. For example, I may want to compress everything before uploading.

Not all of these are needed now, but it would be good to make sure we have some type of idea of the path forward here.

Reply:
Nice summary! I'll create a backlog item for this specific feature because it seems out of scope of this PR, and is additive.